src/mlpack/tests/main_tests/range_search_test.cpp

```diff
@@ -0,0 +1,292 @@
+}
+
+BOOST_AUTO_TEST_CASE(ModelCheck)
+{//check if an output model can be used again for a different usage
+  cout<<"Model Reusability Test"<<endl;
+  arma::mat input_data, query_data;
+  double min_v = 0, max_v = 3;
+  std::string distance_file = "distances.csv";
+  std::string neighbor_file = "neighbors.csv";
+
+  if (!data::Load("iris.csv", input_data))
+    BOOST_FAIL("Unable to load dataset iris.csv!");
+  if (!data::Load("iris_test.csv", query_data))
+    BOOST_FAIL("Unable to load dataset iris_test.csv!");
+
+  SetInputParam("reference", std::move(input_data));
+  SetInputParam("min", min_v);
+  SetInputParam("max", max_v);
+  SetInputParam("distances_file", distance_file);
+  SetInputParam("neighbors_file", neighbor_file);
+  cout<<" 1.Parameters Set"<<endl;
+
+  mlpackMain();
+
+  RSModel* output_model = std::move(CLI::GetParam<RSModel*>("output_model"));
+  CLI::GetSingleton().Parameters()["reference"].wasPassed=false;
+  cout<<" 2.Model Created and copied"<<endl;
+
+  SetInputParam("input_model", std::move(output_model));
+  SetInputParam("query", std::move(query_data));
+
+  mlpackMain();
+
+  cout<<" 3.Model Reinserted"<<endl;
+  if (!( output_model == CLI::GetParam<RSModel*>("output_model") ))
+  {
+    BOOST_FAIL("Models are not Equal");
+  }
+  else
+  {
+    cout<<"Model Checking Test Passed"<<endl<<endl;
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LeafValueTesting)
+{
+  cout<<"Leaf Value Tests"<<endl;
+  //Testing 3 different leaf values - default 20, 15 and 25
+  //Ensure that results match for different leaf values
+  arma::mat x={{0, 3, 3, 4, 3, 1},{4, 4, 4, 5, 5, 2},{0, 1, 2, 2, 3, 3}};
+  std::string distance_file = "distances.csv";
+  std::string neighbors_file = "neighbors.csv";
+  double min_v = 0, max_v = 3;
+  vector<std::vector<int>> neighbor_val = {{}, {2, 3, 4},{1, 3, 4, 5},{1, 2, 4},
+                                         {1, 2, 3},{2}};
+  vector<std::vector<float>> distance_val = {{},{1, 1.73205, 2.23607},
+                                             {1, 1.41421, 1.41421, 3},
+                                             {1.73205, 1.41421, 1.41421},
+                                             {2.23607, 1.41421, 1.41421},{3}};
+  math::Range r(min_v, max_v);
+  vector<vector<size_t>> neighbors, neighbors_temp;
+  vector<vector<double>> distances, distances_temp;
+  vector<int> arr{20, 15, 25};
+  SetInputParam("reference", x);
+  //To prevent warning for lack of definition
+  SetInputParam("min", min_v);
+  SetInputParam("max", max_v);
+  SetInputParam("distances_file", distance_file);
+  SetInputParam("neighbors_file", neighbors_file);
+  SetInputParam("leaf_size", arr[0]);
+  cout<<" Setting Base size for testing :"<<arr[0]<<endl;
+  //Default leaf size is 20
+
+  mlpackMain();
+
+  RSModel* output_model1 = std::move(CLI::GetParam<RSModel*>("output_model"));
+
+  output_model1->Search(r, neighbors, distances);
+
+  bindings::tests::CleanMemory();
+
+  for(size_t i = 1 ; i < arr.size() ; i++)
```

Comment:
I need to think about this one, but changing the leaf size will produce the exact same results for each search.  However, it will produce a different tree.  So, perhaps one thing you could do would be to serialize the trained `RSModel` to a string (by creating a `TextArchive` or something wrapped around a `std::ostringstream` and serializing to that in the usual way specified by boost::serialization documentation), and ensure that each string is different.  That would test that the resulting models contain trees built with different leaf sizes.